Fixed rspec --bisect to sort ids_to_run as the original order

makicamel opened the pull request:
In `RSpec::Core::Bisect::ExampleMinimizer#abort_if_ordering_inconsistent`, minimizer checks the execution order of the devided ids are the same as the original order. So we need to sort them as the original order before run.

For example,
When we have 4 examples: 1.rb, 2.rb, 3.rb, 4.rb and 3.rb is failure order-dependent with 2.rb.
We expect to minimizer reproduce command `rspec 2.rb[1] 3.rb[1]`, but minimizer raises 'The example ordering is inconsistent.'
This is because the ids to run are in the order 4.rb 2.rb (diffrent from the original), so fixed to sort them in the same order as the original error.

pirj commented on lib/rspec/core/bisect/example_minimizer.rb:
```diff
@@ -136,7 +136,7 @@ def non_failing_example_ids
 
         def get_expected_failures_for?(ids)
-          ids_to_run = ids + failed_example_ids
+          ids_to_run = all_example_ids & (ids + failed_example_ids)
```
Nice trick!
If we wouldn’t have to support older Ruby versions, we could have used `intersection`.

What makes me wonder is that for later Ruby versions the method doc explicitly mentions that [the order is preserved](https://ruby-doc.org/core-2.7.0/Array.html#method-i-26), but not for [older ones](https://ruby-doc.org/core-1.8.7/Array.html#method-i-26).
I can’t tell just by looking at the C code if the stable ordering is true for older Ruby, too.

Do you think we have sufficient specs to cover the order?

makicamel replied:
Thank you for your review!

> If we wouldn’t have to support older Ruby versions, we could have used intersection.

Yeah, and I also like `&`, it's very short :)

> Do you think we have sufficient specs to cover the order?

Yes, I think we have sufficient specs:
https://github.com/rspec/rspec-core/blob/v3.13.0/spec/rspec/core/bisect/example_minimizer_spec.rb#L13-L25
This spec fails if Ruby doesn't preserve the left array's order. And this repository tests Legacy Rubys.

makicamel replied:
I found related issue:
https://bugs.ruby-lang.org/issues/7803
https://github.com/ruby/ruby/commit/933ff1d6d73eb43c1c57dc1ce8b25b172d7064bc

It seems to older Ruby has also preserved the order, and document reflected that implementation.

pirj commented:
>> sufficient specs to cover the order?

Do you think it’s possible to cover the 4.rb 2.rb case that you mention in the description in a spec that would be red before the fix?

We usually do it ourselves, but I couldn’t find your real name. May I kindly ask you to add a Changelog entry?

makicamel commented:
> Do you think it’s possible to cover the 4.rb 2.rb case that you mention in the description in a spec that would be red before the fix?

Ahh, I apologize the lack of explanation :pray:
[This spec](https://github.com/rspec/rspec-core/blob/v3.13.0/spec/rspec/core/bisect/example_minimizer_spec.rb#L13-L25) covers the 4.rb 2.rb case that I mentioned in the description. But that spec was green before this fix because `FakeBisectRunner#run` has a tiny mistake to unnecessary sort ids. This is the reason I fixed `FakeBisecrRunner#run` in this PR.
(I'm not good at English, so sorry if I didn't catch your intention :pray:)

And I tried to add a changelog! If I should squash commits, I'll do :)

pirj commented:
If we add the `sort` back accidentally, will any spec fail? Is it possible to write such a spec?

makicamel commented:
I see. 
Because FakeBisectRunner is a test helper, I think it would be difficult to write a spec using a test helper fail when the test helper has a bug.

For example, next spec fails when we add `sort` back accidentally.

```ruby
    it do
      all_ids = %w[ 1.rb[1] 2.rb[1] f.rb[1] 3.rb[1] ]
      dependent_ids = %w[ 2.rb[1] ]
      failure_id = "f.rb[1]"
      patterns = [
        [all_ids, [failure_id]],
        [[failure_id], []],
        [all_ids & (all_ids.slice(all_ids.size/2 .. -1) + [failure_id]), []],
        [all_ids & (all_ids.slice(0 .. all_ids.size/2) + [failure_id]), [failure_id]],
        [all_ids & (all_ids.slice(all_ids.size/2/2 .. all_ids.size/2) + [failure_id]), [failure_id]]
      ]
      allow(RSpec::Core::Bisect::ExampleSetDescriptor).to receive(:new).and_call_original
      patterns.each { |pattern|
        expect(RSpec::Core::Bisect::ExampleSetDescriptor).to receive(:new).with(*pattern)
      }

      fake_runner = FakeBisectRunner.new(
        all_ids,
        [],
        { failure_id => dependent_ids }
      )
      minimizer = new_minimizer(fake_runner)
      minimizer.find_minimal_repro
      expect(minimizer.repro_command_for_currently_needed_ids).to eq("rspec 2.rb[1] f.rb[1]")
    end
```

This spec tests a test helper FakeBisectRunner itself, so I'm not sure this spec should be added.

pirj commented:
Alright. But we've changed how `get_expected_failures_for?` works, which is not a spec helper. Is it possible to safeguard those changes with a spec? This spec would also 
Please don't hesitate to tell if I'm talking a complete nonsense here. I'm reasoning from a generic position - if a change was made to the production code, there should either be:
- a way to reproduce the original issue reliably, and manually test that it fails without the fix and passes with the fix
- a way to reliably cover the same in a form of a (preferably human-readable) spec, both to document what was fixed, and to prevent regressions

Is it possible? Does it make sense? Does it apply in this case?

makicamel commented:
Yeah, sorry for my bad explanation :pray:

First, because spec helper (FakeBisectRunner) has a tiny bug and I fixed it removing unnecessary `sort`. And then, when run [this spec](https://github.com/rspec/rspec-core/blob/v3.13.0/spec/rspec/core/bisect/example_minimizer_spec.rb#L13-L25) with the code removed spec helper's bug, it fails.
Second, now we have a failing spec, I fixed `get_expected_failures_for?`, library code.
So, the bug in the spec helper was getting in the way, but the spec already had a way to reliably reproduce the problem.

For documentation purpose the following spec could be added, but it would overlap with [this spec](https://github.com/rspec/rspec-core/blob/v3.13.0/spec/rspec/core/bisect/example_minimizer_spec.rb#L13-L25).

```ruby
    it 'returns the reproduce command correctly even if there are unrelated ids after the failure id' do
      fake_runner = FakeBisectRunner.new(
        %w[ 1.rb[1] 2.rb[1] 3.rb[1] 4.rb[1] ],
        [],
        { "3.rb[1]" => %w[ 2.rb[1] ] }
      )
      minimizer = new_minimizer(fake_runner)
      minimizer.find_minimal_repro
      expect(minimizer.repro_command_for_currently_needed_ids).to eq("rspec 2.rb[1] 3.rb[1]")
    end
```

makicamel commented:
Thank you! I added an integration spec. 8df0aa697a05485a2d5666b7cbbe4f8a1b33a9f6

JonRowe commented:
Can you rebase this once #3097 is merged / or cherry-pick that commit, theres an issue with our legacy actions needing fixing.

makicamel commented:
@JonRowe 
Thank you for fixing CI! I rebased.

JonRowe commented:
Thanks

JonRowe commented:
Released in 3.13.1 apologies for the delay
